gateway/api/connections/connections.go

```diff
@@ -638,45 +636,203 @@ func GetDatabaseSchemas(c *gin.Context) {
 		}
 	}()
 
-	timeoutCtx, cancelFn := context.WithTimeout(context.Background(), time.Second*50)
+	timeoutCtx, cancelFn := context.WithTimeout(context.Background(), time.Second*30)
 	defer cancelFn()
 	select {
 	case outcome := <-respCh:
 		if outcome.ExitCode != 0 {
-			log.Errorf("failed issuing plain exec: %s, output=%v", outcome.String(), outcome.Output)
-			c.JSON(http.StatusInternalServerError, gin.H{"message": fmt.Sprintf("failed to get schema: %s", outcome.Output)})
+			log.Warnf("failed issuing plain exec: %s, output=%v", outcome.String(), outcome.Output)
+			c.JSON(http.StatusBadRequest, gin.H{"message": fmt.Sprintf("failed to list tables: %s", outcome.Output)})
 			return
 		}
 
-		var schema openapi.ConnectionSchemaResponse
-		var err error
+		response := openapi.TablesResponse{Schemas: []openapi.SchemaInfo{}}
 
 		if currentConnectionType == pb.ConnectionTypeMongoDB {
-			output := cleanMongoOutput(outcome.Output)
-			schema, err = parseMongoDBSchema(output)
+			// Parse MongoDB output
+			tables, err := parseMongoDBTables(outcome.Output)
+			if err != nil {
+				log.Errorf("failed parsing mongo response: %v", err)
+				c.JSON(http.StatusInternalServerError, gin.H{"message": fmt.Sprintf("failed to parse MongoDB response: %v", err)})
+				return
+			}
+			response = tables
 		} else {
-			schema, err = parseSQLSchema(outcome.Output, currentConnectionType)
+			// Parse SQL output
+			tables, err := parseSQLTables(outcome.Output, currentConnectionType)
+			if err != nil {
+				log.Errorf("failed parsing SQL response: %v", err)
+				c.JSON(http.StatusInternalServerError, gin.H{"message": fmt.Sprintf("failed to parse SQL response: %v", err)})
+				return
+			}
+			response = tables
+		}
+
+		c.JSON(http.StatusOK, response)
+
+	case <-timeoutCtx.Done():
+		client.Close()
+		log.Warnf("timeout (30s) obtaining tables for database '%s' using connection '%s'", dbName, conn.Name)
+		c.JSON(http.StatusRequestTimeout, gin.H{
```

Comment:
This status code should be used when the "server didn't receive a request from the client within the expected timeframe". This is not the case here. Change to a 400 status code instead